framework/src/modules/chain/blocks/blocks.js

```diff
@@ -584,6 +705,130 @@ class Blocks extends EventEmitter {
 			}
 		);
 	}
+
+	/**
+	 * Block IDs are the same ~ Blocks are equal
+	 * @param block
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleSameBlockReceived(block) {
+		this.logger.debug('Block already processed', block.id);
+	}
+
+	/**
+	 * Block received is correct
+	 * @param block
+	 * @returns {Promise}
+	 * @private
+	 */
+	_handleGoodBlock(block) {
+		return this._processReceivedBlock(block);
+	}
+
+	/**
+	 * Double forging. Last block stands
+	 * @param block
+	 * @param lastBlock
+	 * @returns {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleDoubleForging(block, lastBlock) {
+		this.logger.warn(
+			'Delegate forging on multiple nodes',
+			block.generatorPublicKey
+		);
+		this.logger.info(
+			`Last block ${lastBlock.id} stands, new block ${block.id} is discarded`
+		);
+		// TODO: Implement Proof of Misbehavior
+	}
+
+	/**
+	 * Tie break: two competing blocks by different delegates at the same height.
+	 * @param lastBlock
+	 * @param newBlock
+	 * @returns {Promise}
+	 * @private
+	 */
+	async _handleDoubleForgingTieBreak(newBlock, lastBlock) {
+		const block = cloneDeep(newBlock);
+		// It mutates the argument
+		const check = await this.blocksVerify.normalizeAndVerify(
+			block,
+			lastBlock,
+			this._lastNBlockIds
+		);
+
+		if (!check.verified) {
+			const errors = convertErrorsToString(check.errors);
+
+			this.logger.error(
+				`Fork Choice Case 4 recovery failed because block ${
+					block.id
+				} verification and normalization failed`,
+				errors
+			);
+			// Return first error from checks
+			throw errors;
+		}
+
+		// If the new block is correctly validated and verified,
+		// last block is deleted and new block is added to the tip of the chain
+		this.logger.info(
+			`Deleting last block with id: ${
+				lastBlock.id
+			} due to Fork Choice Rule Case 4`
+		);
+		const previousLastBlock = cloneDeep(this._lastBlock);
+
+		// Deletes last block and assigns new received block to this._lastBlock
+		await this.recoverChain();
+
+		try {
+			await this._processReceivedBlock(block);
+		} catch (error) {
+			this.logger.warn(
+				`Failed to apply newly received block with id: ${
+					block.id
+				}, restoring previous block ${previousLastBlock.id}`
+			);
+
+			await this._processReceivedBlock(previousLastBlock);
+		}
+	}
+
+	/**
+	 * Move to a different chain
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleMovingToDifferentChain() {
+		// TODO: Move to a different chain.
+		// Determine which method to use to move to a different chain: Block Sync Mechanism or Fast Chain Switching Mechanism
+	}
+
+	/**
+	 * Handle discarded block determined by the fork choice rule
+	 * @param block
+	 * @return {*}
+	 * @private
+	 */
+	// eslint-disable-next-line class-methods-use-this
+	_handleDiscardedBlock(block) {
+		// Discard newly received block
+		return this.logger.warn(
```

Comment:
log does not return anything (ie no return required.
Also, same as above, warn level should not be used within this module